Add unit and e2e tests

khanh-nguyen opened the pull request:
@scottcorgan, I added some unit and e2e tests. It's still WIP but I created a PR to see if you need these tests and if I'm on the right track.

scottcorgan commented:
This looks fantastic. This is perfect timing too. I'm about to do some work on `3.0` and I needed tests to ensure no regressions.

Probably add a travis-ci badge and .travis.yml file too.

Let me know when it's ready to merge.

khanh-nguyen commented:
Great to know that it's helpful. I'll another e2e test for output with fail tests tonight and it will be good for now.

scottcorgan commented:
There will be some new data provided on errors in 3.0, specifically a preview of the specific line in the file of the failing test.

I'll keep the e2e tests in mind when making this change.

khanh-nguyen commented:
I've made some changes. Can you review when you have a chance? Thanks,

scottcorgan commented:
Things are looking very good! A couple things:

* How fragile are the e2e tests? How hard will they be to update with output changes?
* Maybe put the test scripts into their own scripts `test-e2e` and `test-unit` then under `test` you can put ` npm run test-unit && npm run test-e2e`. Makes things cleaner IMO.
* As you are cleaning things up, make sure you attempt to make the test code look similar to the production code. (i.e. using a `var` for every variable definition, etc.)

Thanks again for doing this!

khanh-nguyen commented:
* I've updated `scripts` and style as you said.
* e2e tests follow closely your code logic in `index.js`. So if you change the way you format the output, the tests will have to be updated. But it will be quick I believe.

scottcorgan commented:
Nice!

How soon do you think until the test coverage is complete?

khanh-nguyen commented on index.js:
```diff
@@ -135,10 +135,10 @@ module.exports = function() {
     // Fail number
     if (res.fail.length > 0) {
-      out.push(format.red('  failing:   ' + res.fail.length + '\n'));
+      out.push(format.red('  failing:   ' + res.fail.length) + '\n');
```
@scottcorgan Just in case you didn't notice, I made some small changes here. The problem with formatting the text, including `\n`, is that the color code appears after `\n`. Then, when splitting the output into lines to remove the `duration ...` line, the color code is removed with that line. 
This change doesn't affect the ouput much but make e2e test easier.

scottcorgan replied:
Ya, saw that. Good idea. Made some changes in other modules I have too when I saw that.

khanh-nguyen commented:
Do you want to cover all the cases in [here](https://github.com/substack/tap-parser/tree/master/test)? I won't only have time to look into those details until the weekend. But at first glance, they are very similar to what we already have in our e2e test right now.

scottcorgan commented:
We don't need to cover all of them, just enough that when I do my `3.0` refactor, I'm confident things will work correctly.

I think test coverage for the various things that get parsed, will be enough:

* test titles
* assert titles
* errors
* etc.

khanh-nguyen commented:
OK. I see your point. I'll try to add more work at the end of this week. Is that good? What is your timeline for `3.0`?

khanh-nguyen commented:
* test titles and assert titles are special cases of comments. They are covered in existing unit test and e2e test.
* errors are covered in e2e not-ok test.
* I added `extra` into `not-ok.txt` to be used in e2e test. I also added a unit test for `extra`.
Can't think of other tests to add. If you have anything else in mind, please let me know, @scottcorgan.

scottcorgan commented:
Looks good to merge. If more needs to be done, we can revisit during the refactor. I'll have this merged later today.

scottcorgan commented:
Also, for the sake of records, this references #20

khanh-nguyen commented:
:thumbsup: Cool. Thanks for this useful repo. Keep me posted with your updates and let me know if you need any contribution.

scottcorgan commented:
Looking good. Will publish as `2.2.0`. Will start work on `3.0` this week..
